connman/src/service.c

```diff
@@ -9244,6 +9244,10 @@ static void update_from_network(struct connman_service *service,
 		service->network = connman_network_ref(network);
 		connman_network_autoconnect_changed(service->network,
 							service->autoconnect);
+		__connman_service_network_property_changed(service, "BSSID");
+		__connman_service_network_property_changed(service, "MaxRate");
+		__connman_service_network_property_changed(service, "Frequency");
+		__connman_service_network_property_changed(service, "EncryptionMode");
```

Comment:
I don't think this is the right place for these. These are appended to the DBus struct by `append_wifi_ext_info()` and the struct is then sent when service.c schedules changes about the services to be sent. Those calls to `__connman_service_network_property_changed()` would in this case just send a value update even when the value has not been changed.

If the intention is to get these values updated when the new network for a WiFi service is established I'm pretty sure that this is covered by the scheduled property changes of the services. network.c does send the updates when the values change, at least for `BBSID` and `Frequency` as they can change during the established connection to the network.

In what scenario this issue reproduces and why these are needed? As there is no task to refer to I'm kind of shooting half blindly into the dark with my comments.

Reply:
Your are right, these properties are appended but under normal conditions. Saved wifi `services` are added to connman services list after device start. And after `saved` network adds to already existed `service` we have missing BSSID and other props in mw layer (libconnman-qt, mdm and etc.). Thats why I suggested to explicitly send notifications for these properties when we replace `network` (that was NULL) in existing `saved` `service`. Otherwise these properties always empty (NULL) in mw layer.

Reply:
Since this is the case, thanks for the explanation, I see that this issue happens because the service is not a new, added service, for which all the properties are added by calling [`append_struct()`](https://github.com/sailfishos/connman/blob/master/connman/src/service.c#L3261)b in [`services_append_added_foreach()`](https://github.com/sailfishos/connman/blob/master/connman/src/service.c#L5984). This happens because [`append_struct_service()`](https://github.com/sailfishos/connman/blob/master/connman/src/service.c#L3436) without any callback just adds the path of an existing service, this is called by [`append_struct()`](https://github.com/sailfishos/connman/blob/master/connman/src/service.c#L3261) with `append_dict_properties()` as callback so if the service is in added it will get have the properties updated. The final call here is to `service_list_sort()` which in case of more than one service triggers the sending of changed.

However, this is originally triggered via network.c by calling `__connman_service_create_network()` and if there is no `path` defined the service is treated as a new one, but if the service has path, so the [network comes back](https://github.com/sailfishos/connman/blob/master/connman/src/service.c#L9291) it should every time trigger the `service_schedule_added()` as it is done for the services without `path` set. However, documentation ([manager-api.txt](https://github.com/sailfishos/connman/blob/master/connman/doc/manager-api.txt)) does say that the dict for a service may be empty or contain only changed values. In case of a returning network, these property changed values you have here are those that should be informed. However, this would first send the property changed values and only after the service list, which is done in wrong order and I think these should be included into the main services changed list as properties.

This just gave me an idea of using an `updated` hash table value to `struct services_notify` similar to the `added` with an exception that in [`services_append_added_foreach()`](https://github.com/sailfishos/connman/blob/master/connman/src/service.c#L5984) we'd call another function that simply adds a different callback to the [`append_struct_service()`](https://github.com/sailfishos/connman/blob/master/connman/src/service.c#L3436) to include certain updated values based on the service type. And here with WiFi services it could call the [`append_wifi_ext_info()`](https://github.com/sailfishos/connman/blob/master/connman/src/service.c#L2994). However, [`allow_property_changed()`](https://github.com/sailfishos/connman/blob/master/connman/src/service.c#L6310) suggests using add notify for updates. 

That idea would be then extendable to other service types which may need some additional information added as changed values to the `ServicesChanged` signal by just simply amending the `switch()-case` made for the services that need an update on small range of variables. Any comments on this approach?

Reply:
And it would require creating a new `service_schedule_updated()` which is then called in [`__connman_service_create_from_network()`](https://github.com/sailfishos/connman/blob/master/connman/src/service.c#L9298) when the service has a path set and needs only part of the property changes to be done.

And like with adding a new service to the `ServicesChanged` signal in [`service_append_added_foreach()`](https://github.com/sailfishos/connman/blob/master/connman/src/service.c#L5994) the `updated` could be added by simply calling that `append_struct_service(iter, append_updated_properties, service` where `append_updated_properties()` is the new function to call a specific dict amending function to add the commonly changed parts for the service based on its type. For WiFi, just call `append_wifi_ext_info()`. 

I hope this clarifies what I had in mind.

The fully optimal situation would be that each variable for each service would have its own struct with an additional field `bool updated` which could be used to check if a certain parameter for a service has been changed since the services were last updated to clients listening over D-Bus. But that is beyond the scope here :)

Reply:
Hi, great update. Looking good and I added only some small notes there. mostly about the coding style. It might be even that this could be attempted to be upstreamed, I need to check if there are some contradicting changes or does it suffer from the same symptoms but I guess this would bring value also there.

But first, lets get this completed.

Reply:
Looking good now.